File: VantaSpeech-Android/app/src/main/java/com/vantaspeech/data/network/TranscriptionService.kt
package com.vantaspeech.data.network

import android.content.Context
import android.net.Uri
import com.vantaspeech.data.model.TranscriptionResult
import com.vantaspeech.data.model.TranscriptionState
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.RequestBody.Companion.asRequestBody
import okhttp3.logging.HttpLoggingInterceptor
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import java.io.File
import java.util.concurrent.TimeUnit
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class TranscriptionService @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private var baseUrl: String = "http://localhost:8080/"
    private var api: TranscriptionApi? = null

    fun updateServerUrl(url: String) {
        val normalizedUrl = if (url.endsWith("/")) url else "$url/"
        if (normalizedUrl != baseUrl) {
            baseUrl = normalizedUrl
            api = null // Reset API to recreate with new URL
        }
    }

    private fun getApi(): TranscriptionApi {
        if (api == null) {
            val loggingInterceptor = HttpLoggingInterceptor().apply {
                level = HttpLoggingInterceptor.Level.BASIC
            }

            val client = OkHttpClient.Builder()
                .addInterceptor(loggingInterceptor)
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(5, TimeUnit.MINUTES)
                .writeTimeout(5, TimeUnit.MINUTES)
                .build()

            api = Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(client)
                .addConverterFactory(MoshiConverterFactory.create())
                .build()
                .create(TranscriptionApi::class.java)
        }
        return api!!
    }

    suspend fun transcribe(audioFilePath: String): TranscriptionState = withContext(Dispatchers.IO) {
        try {
            val file = File(audioFilePath)
            if (!file.exists()) {
                return@withContext TranscriptionState.Error("Audio file not found")
            }

            val mimeType = getMimeType(file.extension)
            val requestBody = file.asRequestBody(mimeType.toMediaTypeOrNull())
            val multipartBody = MultipartBody.Part.createFormData(
                "file",
                file.name,
                requestBody
            )

            val response = getApi().transcribe(multipartBody)

            if (response.isSuccessful) {
                response.body()?.let { result ->
                    TranscriptionState.Success(result)
                } ?: TranscriptionState.Error("Empty response from server")
            } else {
                val errorBody = response.errorBody()?.string()
                TranscriptionState.Error(errorBody ?: "Unknown error: ${response.code()}")
            }
        } catch (e: Exception) {
            TranscriptionState.Error(e.message ?: "Transcription failed")
        }
    }

    private fun getMimeType(extension: String): String {
        return when (extension.lowercase()) {
            "m4a" -> "audio/mp4"
            "ogg", "opus" -> "audio/ogg"
            "mp3" -> "audio/mpeg"
            "wav" -> "audio/wav"
            "aac" -> "audio/aac"
            else -> "audio/*"
        }
    }
}

File: VantaSpeech-Windows/VantaSpeech/Services/Network/TranscriptionService.cs
using System.Net.Http.Headers;
using System.Text.Json;
using VantaSpeech.Models;

namespace VantaSpeech.Services.Network;

public interface ITranscriptionService
{
    string BaseUrl { get; set; }
    Task<TranscriptionResult?> TranscribeAsync(string audioFilePath);
}

public class TranscriptionService : ITranscriptionService, IDisposable
{
    private HttpClient _httpClient;
    private string _baseUrl = "http://localhost:8080/";

    public string BaseUrl
    {
        get => _baseUrl;
        set
        {
            var normalizedUrl = value.EndsWith("/") ? value : value + "/";
            if (_baseUrl != normalizedUrl)
            {
                _baseUrl = normalizedUrl;
                ConfigureHttpClient();
            }
        }
    }

    public TranscriptionService()
    {
        _httpClient = new HttpClient();
        ConfigureHttpClient();
    }

    private void ConfigureHttpClient()
    {
        _httpClient.BaseAddress = new Uri(_baseUrl);
        _httpClient.Timeout = TimeSpan.FromMinutes(10);
    }

    public async Task<TranscriptionResult?> TranscribeAsync(string audioFilePath)
    {
        if (!File.Exists(audioFilePath))
        {
            throw new FileNotFoundException("Audio file not found", audioFilePath);
        }

        try
        {
            using var content = new MultipartFormDataContent();
            using var fileStream = File.OpenRead(audioFilePath);
            using var streamContent = new StreamContent(fileStream);

            var mimeType = GetMimeType(Path.GetExtension(audioFilePath));
            streamContent.Headers.ContentType = new MediaTypeHeaderValue(mimeType);

            content.Add(streamContent, "file", Path.GetFileName(audioFilePath));

            var response = await _httpClient.PostAsync("transcribe", content);
            response.EnsureSuccessStatusCode();

            var json = await response.Content.ReadAsStringAsync();
            return JsonSerializer.Deserialize<TranscriptionResult>(json);
        }
        catch (HttpRequestException ex)
        {
            throw new Exception($"Network error: {ex.Message}", ex);
        }
        catch (TaskCanceledException)
        {
            throw new Exception("Request timed out");
        }
    }

    private static string GetMimeType(string extension)
    {
        return extension.ToLowerInvariant() switch
        {
            ".m4a" => "audio/mp4",
            ".ogg" or ".opus" => "audio/ogg",
            ".mp3" => "audio/mpeg",
            ".wav" => "audio/wav",
            ".aac" => "audio/aac",
            _ => "audio/*"
        };
    }

    public void Dispose()
    {
        _httpClient.Dispose();
        GC.SuppressFinalize(this);
    }
}
